[Perf][Frontend]: eliminate api_key and x_request_id headers middleware overhead

Yazan-Sharaya opened the pull request:
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose
The current implementation for the middleware that handles authentication (--api-key) and X-Request-Id headers (--enable-request-id-headers) uses a high level abstraction provided by FastAPI that allows middleware to be written as functions with access to the Request object.
This causes an unacceptable performance hit when the middleware is installed.
This PR re-implements the same functionality using pure ASGI middleware which causes almost zero overhead.
Fixes: #11529

## Test Plan
Create a new test file that verifies that the middleware actually works as expected since currently no tests exist for this.
Benchmark the performance improvement before and after the update.

## Test Result
All the correctness tests pass as expected.
<details>
<summary>Performance Test Results</summary>
The test was conducted using `vllm bench serve` to get an end to end result.
As can be seen, ~10% decrease in throughput is huge for the little computation that is actually done
when enabling both middleware.


## Current implementation

| Configuration | QPS | Average | Difference |
|---------------|-------------------|---------|------------|
| without middleware | 94 - 94 - 96 - 96 - 94 | 94.8 | - |
| --api-key | 90 - 92 - 91 - 90 - 90 | 90.6 | 4.4% |
| --api-key and --enable-request-id-headers | 87 - 86 - 86 - 84 - 84 - 86 | 85.5 | 9.8% |

## Pure ASGI

| Configuration | QPS | Average | Difference |
|---------------|-------------------|---------|------------|
| without middleware | 93 - 95 - 95 - 93 - 93 - 95 | 94.0 | - |
| --api-key | 93 - 95 - 94 - 93 - 93 - 95 | 93.8 | 0.2% |
| --api-key and --enable-request-id-headers | 93 - 93 - 95 - 93 - 92 - 94 | 93.3 | 0.7% |

</details>

## (Optional) Documentation Update
Removed the section in the documentation that warns about the performance penalty of --enable-request-id-headers at high QPS since it's no longer applicable.
Removed the "Caution: " segment in the help text for the parser.add_argument("--enable-request-id-headers") for the same reason as above.

<!--- pyml disable-next-line no-emphasis-as-heading -->

Yazan-Sharaya committed:
perf[frontend]: improve api_token and x_request_id headers middleware performance

Eliminated the overhead of optional middleware by switching to a pure ASGI implementation

Signed-off-by: Yazan-Sharaya <yazan.sharaya.yes@gmail.com>

Yazan-Sharaya committed:
style: fix formatting via pre-commit

Signed-off-by: Yazan-Sharaya <yazan.sharaya.yes@gmail.com>

Yazan-Sharaya committed:
refactor: add type narrowing to satisfy mypy

Signed-off-by: Yazan-Sharaya <yazan.sharaya.yes@gmail.com>

Yazan-Sharaya committed:
refactor: add default argument to `headers.get` to satisfy mypy

Signed-off-by: Yazan-Sharaya <yazan.sharaya.yes@gmail.com>

cjackal commented:
LGTM, especially since now we have proper docstrings for middlewares which can save time for future contributors.

njhill reviewed (commented):
Thanks @Yazan-Sharaya this looks great!

Yazan-Sharaya committed:
Merge branch 'main' into eliminate_middleware_overhead

# Conflicts:
#	docs/serving/openai_compatible_server.md

Yazan-Sharaya committed:
refactor: use method instead of if and f-strings for clarity

Signed-off-by: Yazan-Sharaya <yazan.sharaya.yes@gmail.com>

Yazan-Sharaya commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1062,32 +1063,86 @@ async def validation_exception_handler(_: Request,
+                url_path = URL(scope=scope).path
+                headers = Headers(scope=scope)
+                if root_path and url_path.startswith(root_path):
+                    url_path = url_path[len(root_path):]
```
Changed it to use `removeprefix()` and got rid of the if statement

Yazan-Sharaya commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1062,32 +1063,86 @@ async def validation_exception_handler(_: Request,
+                                            status_code=401)
+                    await response(scope, receive, send)
+                    return
+                await self.app(scope, receive, send)
```
Hmm, I copied the style of `await self.app()` then a bare `return` from starlette's middleware implementation, which is what FastAPI relies on for middleware (and other) functionality under the hood.
Take a look at [this](https://github.com/encode/starlette/blob/master/starlette/middleware/cors.py#L90) or [this](https://github.com/encode/starlette/blob/master/starlette/middleware/authentication.py#L31).

Returning the app/response instead of awaiting it still works correctly but I don't know how changing the  `__call__` from `async` to `sync` would affect the performance.

njhill replied:
It just avoids an async boundary which may yield to the event loop / context switch.

Yazan-Sharaya replied:
I benchmarked it and it indeed leads to slightly better performance. Updated the code for both classes.


| Method            | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Average | Difference |
|---------------------|-------|-------|-------|-------|-------|---------|------------|
| await app/response  | 66.26 (Reqs/s) | 66.68 | 66.43 | 66.21 | 59.29 | 64.974  | -1.59%     |
| return app/response | 66.16 | 65.38 | 66.66 | 65.79 | 66.14 | 66.026  | -          |

njhill commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1062,32 +1063,84 @@ async def validation_exception_handler(_: Request,
+                root_path = scope.get("root_path", "")
+                url_path = URL(scope=scope).path.removeprefix(root_path)
+                headers = Headers(scope=scope)
+                if not url_path.startswith("/v1"):
+                    await self.app(scope, receive, send)
+                    return
+                # Type narrow to satisfy mypy.
+                if isinstance(token, str) and headers.get(
```
Could simplify
```suggestion
                if url_path.startswith("/v1") and isinstance(token, str) and headers.get(
```

Yazan-Sharaya committed:
refactor: combine if statements and slightly improve performance

Merged two consecutive if statements into one for cleaner logic.
Also made a minor change that slightly improves performance.

Signed-off-by: Yazan-Sharaya <yazan.sharaya.yes@gmail.com>

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1062,32 +1063,78 @@ async def validation_exception_handler(_: Request,
-                return JSONResponse(content={"error": "Unauthorized"},
-                                    status_code=401)
-            return await call_next(request)
+        class AuthenticationMiddleware:
```
nit. You can probably move this class outside of this function to reduce indent space.

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -1062,32 +1063,78 @@ async def validation_exception_handler(_: Request,
-            response = await call_next(request)
-            response.headers["X-Request-Id"] = request_id
-            return response
+        class XRequestIdMiddleware:
```
ditto

aarnphm reviewed (approved):
final few items on the formatting style, otherwise this is great. Thanks!

Yazan-Sharaya committed:
refactor: move middleware classes outside `build_app` function

Signed-off-by: Yazan-Sharaya <yazan.sharaya.yes@gmail.com>

njhill reviewed (approved):
Thanks again @Yazan-Sharaya
